### PR TITLE
Exacthash improvements

### PR DESCRIPTION
**Pull request by chris-ha458**
@ChenghaoMou  I really need you to take a look at this.

This is a difficult process to multithread (not impossible, just error prone) especially since the order is important
(I had a shared np array, indexed list, multiple passes with multiple threads in mind. so far saving the hashes as part of a new column "was" the most promising option)

So, after multiple efforts of trying to make this multithreaded/multiprocessed/shared buffers etc
I finally gave up and just streamlined the single threaded approach and removed batching.

batching is usually used to reduce maximum concurrent memory consumption by reducing the working set, and to chunk workloads to multithread.
However huggingface datasets are backed by arrow data structures which automatically handle the working set issue ( it is transparently an mmaped dataset that the arrow backend only loads into memory as necessary).

so I just removed the batching and the speed is now 10 folds faster.
To be honest I cannot see any need for batching in the single threaded version so maybe we can either just remove it and add it back when it seems necessary.

Of course, even though in my personal testing it seemed to automatically manage memory, this may not be true in other datasets (i tried up to with 40GiB millions of data) so let me know if I am wrong on something.

### CODE REVIEW

File: text_dedup/utils/hashfunc.py [hunk 1]
```diff
@@ -43,6 +67,30 @@ def sha1_hash(data: bytes, d: int = 32) -> int:
...
+        The hex hash value.
+
+    Examples
+    --------
+    >>> sha256_hexdigest(b"hello world")
+    'b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9'
+    >>> len(sha256_hexdigest(b"hello world"))
+    64
+    """
+    return sha256(b"hello world").hexdigest()
```
Comment (ChenghaoMou) [inline text_dedup/utils/hashfunc.py:91]: Wrong input.

**Review by ChenghaoMou — COMMENTED** [review]
Thank you, as always.

**Comment by ChenghaoMou** [comment]
Regarding the parallelization part, I tried too. There is no easy way around this since you need the global variable to be shared and simultaneously updated among threads or processes. Python unfortunately adds a lot of the overhead in this regard.

File: text_dedup/utils/hashfunc.py (hunk 1 above)
Comment (chris-ha458) [reply text_dedup/utils/hashfunc.py:91]: Thanks ill put a fix commit.
Hashes are always difficult.

File: text_dedup/exact_hash.py [hunk 2]
```diff
@@ -41,38 +42,47 @@
...
                 lambda _, idx: not flags[idx],
                 with_indices=True,
-                num_proc=os.cpu_count(),
+                num_proc=NUM_PROC,
+                writer_batch_size=args.batch_size,
             )
 
         with timer("Saving"):
             ds.save_to_disk(args.output)
+        ds.cleanup_cache_files()
```
[the remark this answers is not in the document]
Comment (chris-ha458) [reply text_dedup/exact_hash.py:85]: That seems to be the right approach. I'll remove it from here for now.

File: text_dedup/exact_hash.py [hunk 3]
```diff
@@ -41,38 +42,47 @@
...
-            for idx in tqdm(range(0, len(ds), args.batch_size), desc="Processing..."):
-                batch = ds[idx : idx + args.batch_size]
-                for example in tqdm(batch[args.column], leave=False):
-                    h = hash_func(example.encode("utf-8")).hexdigest()
-                    if h in hashes:
-                        flags.append(True)
-                    else:
-                        flags.append(False)
-                        hashes.add(h)
+            # currently processing is done on a single thread.
```
[the remark this answers is not in the document]
Comment (chris-ha458) [reply text_dedup/exact_hash.py:61]: Okay. I understand TB level datasets are always difficult.
I think I will revert for now, but add some documentation regarding batch size
`batching process  itself is a major bottleneck so try batch sizes from the total length of data and work downwards from there`
Maybe I could set it to default to the total size as well.

**Comment by chris-ha458** [comment]
I understand your comments and I'll prepare fixes for each of them.

It might get ugly and I'll squash as necessary but if it gets too ugly for your liking let me know. I'll just open a totally new PR with only the necessary fixes.

Funny you mention big code though since that is one of my motivations to work on this. Feel free to ask for my assistance on that work if you are still working on it. I'd love to contribute to The Stack(is that what you're working on?) as well

File: text_dedup/exact_hash.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chris-ha458) [reply text_dedup/exact_hash.py:90]: I saw that you've done the necessary fixes yourself! Ill mark this as resolved

File: text_dedup/utils/hashfunc.py (hunk 1 above)
Comment (chris-ha458) [reply text_dedup/utils/hashfunc.py]: fixed by [be1010b](https://github.com/ChenghaoMou/text-dedup/pull/37/commits/be1010b5d3e9a94935127015a62f882f02cc5e44)

File: text_dedup/exact_hash.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chris-ha458) [reply text_dedup/exact_hash.py:64]: I resolved this with `[ds.shard()](https://huggingface.co/docs/datasets/v2.14.3/en/package_reference/main_classes#datasets.Dataset.shard)`

  This is fast, efficient, customizable, predictable, reproducible. Basically it is an official tool that represents what huggingface dataset is. if we scale batchsize to be equal to len(dataset) it is almost as efficient as no batching, and scaling the batch_size down exhibits batching behavior WITHOUT the unexplicable delay due to batching.

**Comment by chris-ha458** [comment]
as you can see, I've implemented an efficient alternative to batching with sharding.
many opportunities including multithreaded computation or byte conversion with map is on the table.
However, actually implementing them would take more time and consideration.
I think this is a good place to leave exact_hash for the time being.

**Comment by ChenghaoMou** [comment]
> Funny you mention big code though since that is one of my motivations to work on this. Feel free to ask for my assistance on that work if you are still working on it. I'd love to contribute to The Stack(is that what you're working on?) as well

Yes. A lot of the work you see here originated from BigCode and BigScience. I will reach out on Slack if there is any opportunity to collaborate.

**Review by ChenghaoMou — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**